src/tests/Common/Platform/platformdefines.cpp

```diff
@@ -321,7 +321,7 @@ DWORD TP_GetFullPathName(LPWSTR fileName, DWORD nBufferLength, LPWSTR lpBuffer)
     return GetFullPathNameW(fileName, nBufferLength, lpBuffer, NULL);
 #else
     char nativeFullPath[MAX_PATH];
-    (void)realpath(HackyConvertToSTR(fileName), nativeFullPath);
+    (void)!realpath(HackyConvertToSTR(fileName), nativeFullPath);
```

Comment:
What does the `!` provide?

Reply:
It's to suppress the unused result warning with gcc for the APIs decorated with `__attribute__((warn_unused_result))`. Most people found it unexpected and counter-intuitive: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425; especially when the similar spec of C++17's `[[nodiscard]]` explicitly allow [void-to-cast suppression](https://timsong-cpp.github.io/cppwp/n4659/dcl.attr#nodiscard-2) and gcc abides by it, but with respect to WUR, there is no standard in C so they made this choice to insist of warning with void.

BTW, we use in a few places as-is:
```
src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:    (void)!write(STDERR_FILENO, message, strlen(message));
src/coreclr/pal/src/exception/signal.cpp:                (void)!write(STDERR_FILENO, StackOverflowMessage, sizeof(StackOverflowMessage) - 1);
src/coreclr/vm/encee.cpp:        (void)!alloca(frameIncrement);
src/native/libs/System.Native/pal_process.c:    (void)! pipe2(waitForChildToExecPipe, O_CLOEXEC);
src/native/libs/System.Native/pal_process.c:    (void)! SystemNative_Pipe(waitForChildToExecPipe, PAL_O_CLOEXEC);
```

Reply:
It's to suppress the unused result warning with gcc for the APIs decorated with `__attribute__((warn_unused_result))`. Most people found it unexpected and counter-intuitive: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66425; especially when the similar spec of C++17's `[[nodiscard]]` explicitly allow [cast-to-void suppression](https://timsong-cpp.github.io/cppwp/n4659/dcl.attr#nodiscard-2) and gcc abides by it, but with respect to WUR, there is no standard in C so they made this choice to insist of warning with void.

BTW, we use this in a few places as-is:
```sh
src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp:754:    (void)!write(STDERR_FILENO, message, strlen(message));
src/coreclr/pal/src/exception/signal.cpp:600:                (void)!write(STDERR_FILENO, StackOverflowMessage, sizeof(StackOverflowMessage) - 1);
src/coreclr/vm/encee.cpp:712:        (void)!alloca(frameIncrement);
src/native/libs/System.Native/pal_process.c:307:    (void)! pipe2(waitForChildToExecPipe, O_CLOEXEC);
src/native/libs/System.Native/pal_process.c:309:    (void)! SystemNative_Pipe(waitForChildToExecPipe, PAL_O_CLOEXEC);
```

Reply:
> It's to suppress the unused result warning with gcc for the APIs decorated with __attribute__((warn_unused_result)). 

Interesting. I didn't know that was a thing. I've been using `(void)` for years and that was the indication I didn't care. I agree it is a bit unexpected. Thanks for the link.